cluster/etcd_elector_test.go

```diff
@@ -0,0 +1,99 @@
+// +build kvdb_etcd
+
+package cluster
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"runtime/pprof"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb/kvdb/etcd"
+	"github.com/stretchr/testify/require"
+)
+
+// GuardTimeout implements a test level timeout guard.
+func GuardTimeout(t *testing.T, timeout time.Duration) func() {
+	done := make(chan struct{})
+	go func() {
+		select {
+		case <-time.After(timeout):
+			err := pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+			require.NoError(t, err)
+			panic("test timeout")
+
+		case <-done:
+		}
+	}()
+
+	return func() {
+		close(done)
+	}
+}
+
+// TestEtcdElector tests that two candidates competing for leadership works as
+// expected and that elected leader can resign and allow others to take on.
+func TestEtcdElector(t *testing.T) {
+	guard := GuardTimeout(t, 5*time.Second)
+	defer guard()
+
+	tmpDir, err := ioutil.TempDir("", "etcd")
+	if err != nil {
+		t.Fatalf("unable to create temp dir: %v", err)
+	}
+
+	etcdCfg, cleanup, err := etcd.NewEmbeddedEtcdInstance(tmpDir, 0, 0)
+	require.NoError(t, err)
+	defer cleanup()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const (
+		election = "/election/"
+		id1      = "e1"
+		id2      = "e2"
+	)
+
+	e1, err := newEtcdLeaderElector(
+		ctx, id1, election, etcdCfg,
+	)
+	require.NoError(t, err)
+
+	e2, err := newEtcdLeaderElector(
+		ctx, id2, election, etcdCfg,
+	)
+	require.NoError(t, err)
+
+	var wg sync.WaitGroup
+	ch := make(chan string, 2)
+
+	wg.Add(2)
+	t0 := time.Now()
+
+	go func() {
+		defer wg.Done()
+		time.Sleep(time.Second)
+		require.NoError(t, e1.Campaign(context.Background()))
+		ch <- id1
+	}()
+
+	go func() {
+		defer wg.Done()
+		require.NoError(t, e2.Campaign(context.Background()))
+		ch <- id2
+
+		time.Sleep(2 * time.Second)
+		e2.Resign()
```

Comment:
Not sure I'm following the test properly, but is it _always_ certain that this goroutine will become the leader?

Reply:
Ah I see the first sleep now. IMO a more robust test would be one that doesn't assume any ordering w.r.t which goroutine becomes leader, so it would be something like:
  * Launch both goroutines
  * Wait until one of them becomes leader
  * Make the leader resign 
  * Assert that the other goroutine now becomes leader

Reply:
The `Leader` method should be useful here, along w/ the `Observe` call as well for waiting until something happens re leader.

Reply:
Yes, totally. Made some changes to the tests such that any of the two goroutines can become the leader, resign and then we expect that the other one becomes the next leader. The `Leader` method is indeed very useful for this  so also added to the interface. I'm a bit hesitant to add the `Observe` method since it's a bit misleading as it'll allow us to observe past elections in some cases (see implementation details: https://github.com/etcd-io/etcd/blob/30ce6067da5f68ceef762a1954a6a4c0a648d8bf/client/v3/concurrency/election.go#L178). Not sure if we'll even need it?

Reply:
Ah interesting re `Observe`, that behavior doesn't seem very well documented, I think we can skip it given all we really need is the `Leader` method.